lib/norikra/engine.rb

```diff
@@ -218,6 +218,21 @@ def deregister(query_name)
       end
     end
 
+    def replace(query)
+      info "replacing query", name: query.name, targets: query.targets, expression: query.expression
+      if @queries.select{|q| q.name == query.name }.size == 0 &&
+         @suspended_queries.select{|q| q.name == query.name }.size == 0
+        raise Norikra::ClientError, "query name '#{query.name}' does not exists"
+      end
+
+      if reason = query.invalid?
+        raise Norikra::ClientError, "invalid query '#{query.name}': #{reason}"
+      end
+
+      deregister(query.name)
```

Comment:
These operations should be protected by `@mutex`. Without it, this operation may occur race condition for many requests.